order/rpc_parse.go

```diff
@@ -228,21 +228,43 @@ func ParseRPCBatch(prepareMsg *poolrpc.OrderMatchPrepare) (*Batch,
 		BatchTX:       &wire.MsgTx{},
 	}
 
-	// Parse matched orders.
-	for ourOrderHex, rpcMatchedOrders := range prepareMsg.MatchedOrders {
-		var ourOrder Nonce
-		ourOrderBytes, err := hex.DecodeString(ourOrderHex)
-		if err != nil {
-			return nil, fmt.Errorf("error parsing nonce: %v", err)
-		}
-		copy(ourOrder[:], ourOrderBytes)
-		b.MatchedOrders[ourOrder], err = ParseRPCMatchedOrders(
-			rpcMatchedOrders,
-		)
-		if err != nil {
-			return nil, fmt.Errorf("error parsing matched order: "+
-				"%v", err)
+	// Parse matched orders market by market.
+	for leaseDuration, matchedMarket := range prepareMsg.MatchedMarkets {
+		orders := matchedMarket.MatchedOrders
+		for ourOrderHex, rpcMatchedOrders := range orders {
+			var ourOrder Nonce
+			ourOrderBytes, err := hex.DecodeString(ourOrderHex)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing nonce: "+
+					"%v", err)
+			}
+			copy(ourOrder[:], ourOrderBytes)
+			b.MatchedOrders[ourOrder], err = ParseRPCMatchedOrders(
+				rpcMatchedOrders,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing matched "+
+					"order: %v", err)
+			}
+
+			// It is imperative that all matched orders have the
+			// correct lease duration. Otherwise assumptions in the
+			// batch verifier won't be correct. That's why we
+			// validate this here already.
+			for _, matchedOrder := range b.MatchedOrders[ourOrder] {
```

Comment:
Why validate this here (when _parsing_ the batch) vs later in the validation logic?

Reply:
I guess the parsing is part of validation in a sense, as it can cause batch validation failures. Also with this placement, we can catch an invalid batch earlier in the processing cycle.

Reply:
To keep the diff reasonably small, we discard the `duration->orders` mapping after verifying that all orders within a bucket have the correct duration. It's not exactly parsing logic. So we could thread it through to the batch verifier. But as you say, like this we catch format inconsistencies earlier.